pv211_utils/evaluation_metrics.py

```diff
@@ -0,0 +1,405 @@
+        self.k = k
+        self.__class__._CURRENT_INSTANCE = self
+
+        if num_processes == 1:
+            for query in list(queries.values()):
+                map_score += self.__class__._calc_average_precision(query)
+        else:
+            with get_context("fork").Pool(processes=num_processes) as process_pool:
+                for precision in process_pool.imap(self.__class__._calc_average_precision, list(queries.values())):
+                    map_score += precision
+
+        map_score /= len(queries)
+        self.__class__._CURRENT_INSTANCE = None
+        return map_score
+
+
+def _calc_ndcg(system: IRSystemBase, judgements: Set, k: int,
+               query: QueryBase) -> float:
+    num_relevant = 0
+    dcg = 0.0
+    current_rank = 1
+
+    for document in system.search(query):
+        if current_rank > k:
+            break
+        if (query.query_id, document.document_id) in judgements:
+            num_relevant += 1
+            dcg += 1 / log2(current_rank + 1)
+        current_rank += 1
+
+    # max to avoid division by 0
+    idcg = max(sum([1 / log2(i + 1) for i in range(1, num_relevant + 1)]), 1)
+
+    return dcg / idcg
+
+
+def _calc_bpref(system: IRSystemBase, judgements: Set, k: int,
+                query: QueryBase) -> float:
+    num_relevant = 0
+    relevant_doc_ranks = []
+    current_rank = 1
+    bpref = 0.0
+
+    for document in system.search(query):
+        if current_rank > k:
+            break
+        if (query.query_id, document.document_id) in judgements:
+            num_relevant += 1
+            relevant_doc_ranks.append(current_rank)
+        current_rank += 1
+
+    for i, rank in enumerate(relevant_doc_ranks, start=1):
+        bpref += 1 - min(rank - i, num_relevant) / num_relevant
+
+    return (bpref / num_relevant) if num_relevant > 0 else 0
+
+
+def mean_average_precision(system: IRSystemBase, queries: OrderedDict,
+                           judgements: Set[JudgementBase],
+                           k: int, num_processes: int) -> float:
+    """Evaluate system for given queries and judgements with mean average precision
+    metric. Where first k documents will be used in evaluation.
+
+    Arguments
+    ---------
+    system : IRSystemBase
+        System to be evaluated.
+    queries : OrderedDict
+        Queries to be searched.
+    judgements : Set[JudgementBase]
+        Judgements.
+    k : int
+        Parameter defining evaluation depth.
+    num_processes : int
+        Parallelization parameter defining number of processes to be used to run the evaluation.
+
+    Returns
+    -------
+    float
+        Mean average precision score from interval [0, 1].
+    """
+    map_score = 0.0
+
+    if num_processes == 1:
+        for query in list(queries.values()):
+            map_score += _calc_average_precision(system, _judgements_obj_to_id(judgements), k, query)
+    else:
+        worker_avg_precision = partial(_calc_average_precision, system,
```

Comment:
Pickling a system sounds easy to break or, worse, to slow it down to a point where it would be faster to evaluate with one worker when the system references a large object. Let's save the current system to a `_CURRENT_SYSTEM` global if we are keeping `mean_average_precision` rather than `calc_map`?

Reply:
I haven't thought of using a global variable, seems to work. (I will change the other eval metrics to do it like that later.)